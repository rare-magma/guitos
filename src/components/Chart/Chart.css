.recharts-default-tooltip,
.recharts-tooltip-wrapper {
  background-color: var(--lightbgcolor);
  border: 1px solid var(--comment);
}
