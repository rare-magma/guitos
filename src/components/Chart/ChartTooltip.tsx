import Big from "big.js";
import { Col, Container, Row } from "react-bootstrap";
import { useConfig } from "../../context/ConfigContext";
import { intlFormat, roundBig } from "../../utils";

interface ChartTooltipProps {
  active?: boolean;
  payload?: {
    name: string;
    payload?: {
      id: string;
      item: string;
      name: string;
      type: string;
      value: number;
    };
    value: number;
    unit: string;
  }[];
  label?: string;
  key1?: string | undefined;
  key2?: string | undefined;
}

function ChartTooltip({
  active,
  payload,
  label,
  key1,
  key2,
}: ChartTooltipProps) {
  const { intlConfig } = useConfig();
  const currency = intlConfig?.currency;
  const showTooltip = active && payload?.length && currency;
  const filteredItemName = payload?.length && payload[0].payload?.item;
  const showMultipleLegends = showTooltip && payload.length > 1;

  const item1Value =
    showTooltip && intlFormat(roundBig(Big(payload[0].value), 2), currency);
  const item2Value =
    showMultipleLegends &&
    intlFormat(roundBig(Big(payload[1].value), 2), currency);

  return showTooltip ? (
    <Container className="m-2">
      <Row>{label}</Row>
      {showMultipleLegends ? (
        <>
          <Row className="me-1">
            <Col>{`${key1 ?? ""}:`}</Col>
            <Col className="text-end fixed-width-font">{item1Value}</Col>
          </Row>
          <Row className="me-1 flex-md-nowrap">
            <Col>{`${key2 ?? ""}:`}</Col>
            <Col className="text-end fixed-width-font col-md-auto">
              {item2Value}
            </Col>
          </Row>
        </>
      ) : (
        <Row className="me-1">
          <Col>{`${key1 ?? filteredItemName}:`}</Col>
          {key1 === "goal" ? (
            <Col className="text-end fixed-width-font">{`${payload[0].value}%`}</Col>
          ) : (
            <Col className="text-end fixed-width-font">{item1Value}</Col>
          )}
        </Row>
      )}
    </Container>
  ) : null;
}

export default ChartTooltip;
