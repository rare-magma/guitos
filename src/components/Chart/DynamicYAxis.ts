/* eslint-disable eqeqeq */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import { useCallback, useMemo, useState } from "react";

const RECHART_CERTESIAN_AXIS_TICK_VALUE_SELECTOR = `.recharts-cartesian-axis-tick-value[orientation="left"],
.recharts-cartesian-axis-tick-value[orientation="right"]`;

interface Props {
  yAxisWidthModifier?: (width: number) => number;
}

interface ReturnValues {
  yAxisWidth: undefined | number;
  setChartRef: (chartRef: any) => void;
}

function useDynamicYAxisWidth(props: void | Props): ReturnValues {
  const { yAxisWidthModifier } = props || {};
  const [yAxisWidthState, setYAxisWidthState] = useState(undefined);

  const setChartRef = useCallback(
    (chartRef: any) => {
      if (chartRef?.container != null) {
        const tickValueElements = chartRef.container.querySelectorAll(
          RECHART_CERTESIAN_AXIS_TICK_VALUE_SELECTOR,
        );
        const highestWidth = [...tickValueElements]
          .map((el) => {
            const boundingRect = el.getBoundingClientRect();
            if (boundingRect?.width != null) {
              return boundingRect.width;
            }
            return 0;
          })
          .reduce((accumulator, value) => {
            if (accumulator < value) {
              return value;
            }
            return accumulator;
          }, 0);
        setYAxisWidthState(highestWidth);
      }
    },
    [setYAxisWidthState],
  );

  const yAxisWidth = useMemo(() => {
    if (yAxisWidthModifier != null && yAxisWidthState != null) {
      return yAxisWidthModifier(yAxisWidthState);
    }
    return yAxisWidthState;
  }, [yAxisWidthModifier, yAxisWidthState]);

  return {
    yAxisWidth,
    setChartRef,
  };
}

export default useDynamicYAxisWidth;
