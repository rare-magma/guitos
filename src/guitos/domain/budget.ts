import { immerable } from "immer";
import type { BudgetItem } from "./budgetItem";
import type { Expenses } from "./expenses";
import type { Incomes } from "./incomes";
import type { Stats } from "./stats";
import { Uuid } from "./uuid";

export class Budget {
  id: Uuid;
  name: string;
  expenses: Expenses;
  incomes: Incomes;
  stats: Stats;

  [immerable] = true;

  constructor(
    id: Uuid,
    name: string,
    expenses: Expenses,
    incomes: Incomes,
    stats: Stats,
  ) {
    this.id = id;
    this.name = name;
    this.expenses = expenses;
    this.incomes = incomes;
    this.stats = stats;
  }

  static create(date?: string, goal?: number): Budget {
    const newId = Uuid.random();
    const year = new Date().getFullYear();
    const newBudget = new Budget(
      newId,
      date ?? `${year}-${newId.toString().slice(0, 8)}`,
      {
        items: [{ id: 1, name: "", value: 0 }],
        total: 0,
      },
      {
        items: [{ id: 1, name: "", value: 0 }],
        total: 0,
      },
      {
        available: 0,
        withGoal: 0,
        saved: 0,
        goal: goal ?? 10,
        reserves: 0,
      },
    );

    return newBudget;
  }

  static createEmpty(name: string): Budget {
    const emptyExpenses: BudgetItem[] = [];
    const emptyIncomes: BudgetItem[] = [];
    const emptyBudget: Budget = {
      ...Budget.create(name, 0),
      expenses: {
        items: emptyExpenses,
        total: 0,
      },
      incomes: {
        items: emptyIncomes,
        total: 0,
      },
    };
    return emptyBudget;
  }

  static clone(budget: Budget): Budget {
    const clonedBudget: Budget = {
      ...budget,
      id: Uuid.random(),
      name: `${budget.name}-clone`,
    };
    return clonedBudget;
  }

  static toSafeFormat(budget: Budget) {
    return {
      ...budget,
      id: budget.id.toString(),
    };
  }
}
