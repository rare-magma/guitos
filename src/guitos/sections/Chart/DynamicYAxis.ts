import { useCallback, useMemo, useState } from "react";

const RECHART_CARTESIAN_AXIS_TICK_VALUE_SELECTOR = `.recharts-cartesian-axis-tick-value[orientation="left"],
.recharts-cartesian-axis-tick-value[orientation="right"]`;

interface Props {
  yAxisWidthModifier?: (width: number) => number;
}

interface ReturnValues {
  yAxisWidth: undefined | number;
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  setChartRef: (chartRef: any) => void;
}

// biome-ignore lint/suspicious/noConfusingVoidType: <explanation>
export function useDynamicYAxisWidth(props: void | Props): ReturnValues {
  const { yAxisWidthModifier } = props || {};
  const [yAxisWidthState, setYAxisWidthState] = useState(undefined);

  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  const setChartRef = useCallback((chartRef: any) => {
    if (chartRef?.container != null) {
      const tickValueElements = chartRef.container.querySelectorAll(
        RECHART_CARTESIAN_AXIS_TICK_VALUE_SELECTOR,
      );
      const highestWidth = [...tickValueElements]
        .map((el) => {
          const boundingRect = el.getBoundingClientRect();
          if (boundingRect?.width != null) {
            return boundingRect.width;
          }
          return 0;
        })
        .reduce((accumulator, value) => {
          if (accumulator < value) {
            return value;
          }
          return accumulator;
        }, 0);
      setYAxisWidthState(highestWidth);
    }
  }, []);

  const yAxisWidth = useMemo(() => {
    if (yAxisWidthModifier != null && yAxisWidthState != null) {
      return yAxisWidthModifier(yAxisWidthState);
    }
    return yAxisWidthState;
  }, [yAxisWidthModifier, yAxisWidthState]);

  return {
    yAxisWidth,
    setChartRef,
  };
}
