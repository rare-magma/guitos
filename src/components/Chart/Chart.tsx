import { Card, Col, Form, InputGroup, Row } from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import { BsPercent } from "react-icons/bs";
import {
  Area,
  AreaChart,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from "recharts";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { intlFormat, median } from "../../utils";
import { FilteredItem } from "../ChartsPage/ChartsPage";
import "./Chart.css";
import { ChartTooltip } from "./ChartTooltip";
import { useDynamicYAxisWidth } from "./DynamicYAxis";

interface ChartProps {
  header: string;
  tooltipKey1?: string;
  tooltipKey2?: string;
  areaDataKey1: string;
  areaDataKey2?: string;
  areaStroke1: string;
  areaFill1: string;
  areaStroke2?: string;
  areaFill2?: string;
  legend1: string;
  legendValues1: number[];
  legend2?: string;
  legendValues2?: number[];
  unit?: string;
  filteredData?: FilteredItem[];
}

const horizonalRatio = 3.4;
const verticalRatio = 1.6;

export function Chart({
  header,
  tooltipKey1,
  tooltipKey2,
  areaDataKey1,
  areaDataKey2,
  areaStroke1,
  areaStroke2,
  areaFill1,
  areaFill2,
  legend1,
  legend2,
  legendValues1,
  legendValues2,
  unit,
  filteredData,
}: ChartProps) {
  const { budgetList } = useBudget();
  const { intlConfig } = useConfig();

  const showSecondArea = areaDataKey2 && areaStroke2 && areaFill2;
  const isGoalChart = tooltipKey1 === "goal";
  const isVerticalScreen = window.innerWidth < window.innerHeight;
  const chartData =
    filteredData ?? budgetList?.sort((a, b) => a.name.localeCompare(b.name));

  const { yAxisWidth, setChartRef } = useDynamicYAxisWidth({
    yAxisWidthModifier: (x) => x + 10,
  });

  function medianLabelGroup(legend: string, values: number[]) {
    return (
      <InputGroup size="sm" className="mb-1">
        <InputGroup.Text>{legend}</InputGroup.Text>
        <CurrencyInput
          disabled
          className="text-end form-control fixed-width-font"
          aria-label={legend}
          intlConfig={intlConfig}
          defaultValue={median(values)}
        />
      </InputGroup>
    );
  }

  function tickFormatter(value: number) {
    return (
      (intlConfig?.currency && intlFormat(value, intlConfig.currency)) ?? ""
    );
  }

  return (
    <Card className="stat-card mt-3 d-flex flex-grow-1">
      <Card.Header className="stat-card-header">{header}</Card.Header>
      <Card.Body>
        <ResponsiveContainer
          width="100%"
          aspect={isVerticalScreen ? verticalRatio : horizonalRatio}
        >
          <AreaChart
            data={chartData}
            ref={setChartRef}
            margin={{
              top: 10,
              right: 0,
              bottom: 0,
            }}
          >
            <XAxis
              stroke="var(--textcolor)"
              dataKey="name"
              minTickGap={10}
              tick={{ transform: "translate(0, 10)" }}
            />
            {unit ? (
              <YAxis
                stroke="var(--textcolor)"
                style={{ fontFamily: "ui-monospace, monospace" }}
                width={yAxisWidth}
                unit={unit}
              />
            ) : (
              <YAxis
                stroke="var(--textcolor)"
                style={{ fontFamily: "ui-monospace, monospace" }}
                width={yAxisWidth}
                tickFormatter={tickFormatter}
              />
            )}
            <Tooltip
              allowEscapeViewBox={{ x: false, y: false }}
              position={{ y: 0 }}
              content={<ChartTooltip key1={tooltipKey1} key2={tooltipKey2} />}
              contentStyle={{ backgroundColor: "var(--bgcolor)" }}
              itemStyle={{ color: "var(--textcolor)" }}
            />
            <Area
              type="monotone"
              dataKey={areaDataKey1}
              stroke={`var(--${areaStroke1})`}
              fill={`var(--${areaFill1})`}
              isAnimationActive={false}
            />
            {showSecondArea && (
              <Area
                type="monotone"
                dataKey={areaDataKey2}
                stroke={`var(--${areaStroke2})`}
                fill={`var(--${areaFill2})`}
                isAnimationActive={false}
              />
            )}
          </AreaChart>
        </ResponsiveContainer>
      </Card.Body>
      <Card.Footer className="stat-card-footer">
        <Row>
          {isGoalChart && (
            <InputGroup size="sm" className="mb-1">
              <InputGroup.Text>{legend1}</InputGroup.Text>
              <Form.Control
                className="text-end fixed-width-font"
                aria-label={"median"}
                defaultValue={legendValues1 && median(legendValues1)}
                type="number"
                disabled
              />
              <InputGroup.Text>
                <BsPercent aria-hidden />
              </InputGroup.Text>
            </InputGroup>
          )}
          {!isGoalChart &&
            !legendValues2 &&
            medianLabelGroup(legend1, legendValues1)}
          {!isGoalChart && legendValues2 && (
            <Col lg={6}>{medianLabelGroup(legend1, legendValues1)}</Col>
          )}
          {legend2 && legendValues2 && (
            <Col>{medianLabelGroup(legend2, legendValues2)}</Col>
          )}
        </Row>
      </Card.Footer>
    </Card>
  );
}
