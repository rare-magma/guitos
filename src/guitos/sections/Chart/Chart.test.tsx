import { BudgetMother } from "@guitos/domain/budget.mother";
import { Chart } from "@guitos/sections/Chart/Chart";
import { render, screen } from "@testing-library/react";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

describe("Chart", () => {
  const comp = (
    <Chart
      header={"chart header"}
      tooltipKey1={"tooltipKey1"}
      areaDataKey1={"revenue"}
      areaStroke1={"highlight"}
      areaFill1={"highlight"}
      legend1={"median revenue"}
      legendValues1={BudgetMother.testBudgetList().map((b) => {
        return b.incomes.total;
      })}
    />
  );

  beforeEach(() => {
    //@ts-expect-error
    window.ResizeObserver = undefined;
    window.ResizeObserver = vi.fn().mockImplementation(() => ({
      observe: vi.fn(),
      unobserve: vi.fn(),
      disconnect: vi.fn(),
    }));
  });

  afterEach(() => {
    window.ResizeObserver = ResizeObserver;
    vi.restoreAllMocks();
  });

  it("matches snapshot", () => {
    render(comp);
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    render(comp);
    expect(screen.getByText("chart header")).toBeInTheDocument();
    expect(screen.getByText("median revenue")).toBeInTheDocument();
    expect(screen.getByDisplayValue("$200")).toBeInTheDocument();
  });
});
