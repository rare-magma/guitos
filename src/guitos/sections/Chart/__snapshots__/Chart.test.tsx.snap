// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Chart > matches snapshot 1`] = `
<Chart
  areaDataKey1="revenue"
  areaFill1="highlight"
  areaStroke1="highlight"
  header="chart header"
  legend1="median revenue"
  legendValues1={
    [
      100,
      200,
      342883.86,
    ]
  }
  tooltipKey1="tooltipKey1"
/>
`;
